File: src_backup/main/java/com/myudog/myulib/internal/monitor/VFXMonitor.java
package com.myudog.myulib.internal.monitor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.concurrent.atomic.AtomicInteger;
public final class VFXMonitor {
    private static final Logger logger = LoggerFactory.getLogger("MyuVFX-Monitor");
    private static final AtomicInteger particleCount = new AtomicInteger(0);
    public static volatile int maxParticlesPerSecond = 10000;
    private static volatile long lastResetTime = System.currentTimeMillis();
    private VFXMonitor() {
    }
    public static boolean requestSpawn(int amount) {
        checkReset();
        if (particleCount.get() + amount > maxParticlesPerSecond) {
            return false;
        }
        particleCount.addAndGet(amount);
        return true;
    }
    private static void checkReset() {
        long now = System.currentTimeMillis();
        if (now - lastResetTime > 1000L) {
            if (particleCount.get() > 0) {
                logger.debug("Current VFX Load: {} particles/sec", particleCount.get());
            }
            particleCount.set(0);
            lastResetTime = now;
        }
    }
    public static int getCurrentTPSLoad() {
        return particleCount.get();
    }
}

File: src/main/kotlin/com/myudog/myulib/internal/monitor/VFXMonitor.kt
package com.myudog.myulib.internal.monitor

import org.slf4j.LoggerFactory
import java.util.concurrent.atomic.AtomicInteger

internal object VFXMonitor {
    private val logger = LoggerFactory.getLogger("MyuVFX-Monitor")
    private val particleCount = AtomicInteger(0)

    // 安全閾值：每秒最大粒子數 (20 ticks 總計)
    var maxParticlesPerSecond = 10000
    private var lastResetTime = System.currentTimeMillis()

    /**
     * 嘗試申請生成粒子，若超過閾值則回傳 false
     */
    fun requestSpawn(amount: Int): Boolean {
        checkReset()
        if (particleCount.get() + amount > maxParticlesPerSecond) {
            return false
        }
        particleCount.addAndGet(amount)
        return true
    }

    private fun checkReset() {
        val now = System.currentTimeMillis()
        if (now - lastResetTime > 1000) {
            // 每秒回報一次數據（可選）
            if (particleCount.get() > 0) {
                // logger.info("Current VFX Load: ${particleCount.get()} particles/sec")
            }
            particleCount.set(0)
            lastResetTime = now
        }
    }

    fun getCurrentTPSLoad(): Int = particleCount.get()
}